dirty_cat/datetime_encoder.py

```diff
@@ -0,0 +1,171 @@
+        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
+            raise ValueError(
+                f'"extract_until" should be one of ["year", "month", "day", "hour", "minute", '
+                f'"second", "millisecond", "microsecond", "nanosecond"], got {self.extract_until}.'
+            )
+
+    def _extract_from_date(self, date_series, feature):
+        if feature == "year":
+            return pd.DatetimeIndex(date_series).year.to_numpy()
+        elif feature == "month":
+            return pd.DatetimeIndex(date_series).month.to_numpy()
+        elif feature == "day":
+            return pd.DatetimeIndex(date_series).day.to_numpy()
+        elif feature == "hour":
+            return pd.DatetimeIndex(date_series).hour.to_numpy()
+        elif feature == "minute":
+            return pd.DatetimeIndex(date_series).minute.to_numpy()
+        elif feature == "second":
+            return pd.DatetimeIndex(date_series).second.to_numpy()
+        elif feature == "millisecond":
+            return pd.DatetimeIndex(date_series).millisecond.to_numpy()
+        elif feature == "microsecond":
+            return pd.DatetimeIndex(date_series).microsecond.to_numpy()
+        elif feature == "nanosecond":
+            return pd.DatetimeIndex(date_series).nanosecond.to_numpy()
+        elif feature == "dayofweek":
+            return pd.DatetimeIndex(date_series).dayofweek.to_numpy()
+        elif feature == "holiday":
+            # Create an indicator for holidays
+            cal = calendar()
+            holidays = cal.holidays(start=date_series.min(), end=date_series.max())
+            return np.isin(date_series, holidays).astype(int)
+        elif feature == "other":
+            # Gather all the variables below the extract_until into one numerical variable
+            res = (pd.to_datetime(date_series) - pd.to_datetime(pd.DatetimeIndex(date_series).floor(
+                self.word_to_alias[self.extract_until]))).to_numpy()
+            # Convert to the extract_until unit (e.g if I extract until "minute", then convert to minutes)
+            return res / pd.to_timedelta(1, self.word_to_alias[self.extract_until])
+
+    def fit(self, X, y=None):
+        """
+        Fit the DatetimeEncoder to X. In practice, just stores which extracted features
+        are not constant.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_features]
+            Data where each column is a datetime feature.
+        Returns
+        -------
+        self
+            Fitted DatetimeEncoder instance.
+        """
+        self._validate_keywords()
+        self.to_extract_full = self.to_extract[:self.to_extract.index(self.extract_until) + 1]
+        self.to_extract_full.append("other")
+        if isinstance(X, pd.DataFrame):
+            self.colnames = X.columns
+        else:
+            self.colnames = None
```

Comment:
Given that `self.colnames` is determined from the data, it should end with an underscore: `self.colnames_`.

In addition, we prefer to separate words with underscores (to make them easier to read): `self.col_names_`